modules/core/src/main/scala/doobie/hi/connection.scala

```diff
@@ -436,20 +454,27 @@ object connection {
       rsc: ResultSetConcurrency,
       rsh: Holdability
   )(k: PreparedStatementIO[A]): ConnectionIO[A] =
-    IFC.prepareStatement(sql, rst.toInt, rsc.toInt, rsh.toInt).bracket(s => IFC.embed(s, k))(s =>
-      IFC.embed(s, IFPS.close))
+    IFC.prepareStatement(sql, rst.toInt, rsc.toInt, rsh.toInt).flatMap(
+      s => WeakAsyncConnectionIO.cancelable(IFC.embed(s, k).attempt.flatTap(_ => IFC.embed(s, IFPS.close)).rethrow, IFC.embed(s, IFPS.close))
+    )
 
   /** @group Prepared Statements */
   def prepareStatement[A](sql: String, agk: AutoGeneratedKeys)(k: PreparedStatementIO[A]): ConnectionIO[A] =
-    IFC.prepareStatement(sql, agk.toInt).bracket(s => IFC.embed(s, k))(s => IFC.embed(s, IFPS.close))
+    IFC.prepareStatement(sql, agk.toInt).flatMap(
+      s => WeakAsyncConnectionIO.cancelable(IFC.embed(s, k).attempt.flatTap(_ => IFC.embed(s, IFPS.close)).rethrow, IFC.embed(s, IFPS.close))
```

Comment:
`.attempt.flatTap(_ => ...).rethrow` is `.guarantee(...)

Reply:
I'm guessing the intention here is to avoid calling ps.close another time when there's a cancellation?

Reply:
Thanks! I did it this way to prevent double calling of IFPS.close in case of cancelation

```
guaranteeCase {
            case Outcome.Canceled() => WeakAsyncConnectionIO.unit
            case _ => IFC.embed(ps, IFPS.close)
          }
```